Archives/ClmFile.cpp

```diff
@@ -463,97 +458,94 @@ namespace Archives
-		if (WriteFile(outFile, waveFormat, sizeof(WaveFormatEx), &numBytes, NULL) == 0) return false;
-		// Write the unknown bytes
-		if (WriteFile(outFile, m_Unknown, 6, &numBytes, NULL) == 0) return false;
-		// Write the number of internal files
-		if (WriteFile(outFile, &m_NumberOfPackedFiles, 4, &numBytes, NULL) == 0) return false;
-		*/
-
-		// Prepare the index entries
-		offset = sizeof(header) + numFilesToPack * sizeof(IndexEntry);
-		for (i = 0; i < numFilesToPack; i++)
-		{
-			// Copy the filename into the entry
-			strncpy((char*)&entry[i].fileName, internalNames[i].c_str(), 8);
-			// Set the offset of the file
-			entry[i].dataOffset = offset;
-			offset += entry[i].dataLength;
+		unsigned long numBytes;
+		if (WriteFile(outFile, &header, sizeof(header), &numBytes, nullptr) == 0) {
+			return false;
 		}
 
-		// Write the index
-		if (WriteFile(outFile, entry, numFilesToPack * sizeof(IndexEntry), &numBytes, nullptr) == 0)
-		{
+		PrepareIndex(sizeof(header), internalNames, entry);
+		
+		if (WriteFile(outFile, entry, header.packedFilesCount * sizeof(IndexEntry), &numBytes, nullptr) == 0) {
 			// Error writing index table
 			delete[] entry;		// Release the memory for the index table
 			return false;
 		}
 
 		// Copy the files into the volume
-		for (i = 0; i < numFilesToPack; i++)
+		unsigned long numBytesRead;
+		std::array<char, CLM_WRITE_SIZE> buffer;
+		for (int i = 0; i < header.packedFilesCount; i++)
 		{
-			offset = 0;
-			do
-			{
-				numBytes = CLM_WRITE_SIZE;
-				if (offset + numBytes > entry[i].dataLength) numBytes = entry[i].dataLength - offset;
-				// Read the input file
-				if (ReadFile(fileHandle[i], buff, numBytes, &numBytesRead, nullptr) == 0)
-				{
-					// Error reading input file
-					delete[] entry;		// Release the memory for the index table
-					return false;
-				}
-				offset += numBytesRead;
-				// Write the data to the output file
-				if (WriteFile(outFile, buff, numBytesRead, &numBytes, nullptr) == 0)
-				{
-					// Error writing output file
-					delete[] entry;		// Release the memory for the index table
-					return false;
-				}
-			} while (numBytesRead);
+			if (!PackFile(outFile, buffer, numBytesRead, entry[i], fileHandle[i])) {
+				// Error reading input file
+				delete[] entry;	// Release the memory for the index table
+				return false;
+			}
 		}
 
 		return true;
 	}
 
+	void ClmFile::PrepareIndex(int headerSize, const std::vector<std::string>& internalNames, IndexEntry* entry)
+	{
+		int offset = headerSize + internalNames.size() * sizeof(IndexEntry);
+		for (size_t i = 0; i < internalNames.size(); i++)
+		{
+			// Copy the filename into the entry
+			strncpy((char*)&entry[i].fileName, internalNames[i].c_str(), 8);
+			// Set the offset of the file
+			entry[i].dataOffset = offset;
+			offset += entry[i].dataLength;
+		}
+	}
+
+	bool ClmFile::PackFile(HANDLE outFile, std::array<char, CLM_WRITE_SIZE>& buffer, unsigned long& numBytesRead, const IndexEntry& entry, HANDLE fileHandle)
```

Comment:
I'm wondering if this method should take the buffer as a parameter, or just declare it locally. There may be some good reasons either way.